### PR TITLE
feat: add callbacks during training models

### PR DESCRIPTION
**Pull request by hasibzunair**
Adds ReduceLr and EarlyStop callbacks.

### CODE REVIEW

File: decavision/model_training/tfrecords_image_classifier.py [hunk 1]
```diff
@@ -310,14 +310,28 @@ def fit(self, save_model=None, export_model=None, min_accuracy=None,
             callbacks = [tensorboard_callback]
 
         # if we want to stop training when no sufficient improvement in accuracy has been achieved
-        if min_accuracy:
+        if min_accuracy is not None:
```
Comment (yangobeil) [inline decavision/model_training/tfrecords_image_classifier.py]: Actually I think it was ok like it was before.

File: decavision/model_training/tfrecords_image_classifier.py [hunk 2]
```diff
@@ -310,14 +310,28 @@ def fit(self, save_model=None, export_model=None, min_accuracy=None,
...
             early_stop = tf.keras.callbacks.EarlyStopping(monitor='val_sparse_categorical_accuracy',
                                                           min_delta=min_accuracy)
             if not callbacks:
                 callbacks = [early_stop]
             else:
                 callbacks.append(early_stop)
 
+        # use reduce learning rate and early stopping callbacks
+        else:
+            early_stop = tf.keras.callbacks.EarlyStopping(monitor='val_sparse_categorical_accuracy',
```
Comment (yangobeil) [inline decavision/model_training/tfrecords_image_classifier.py:321]: Here my suggestion would be make the patience a parameter so people can decide it, and if it's 0 then not use earlystopping. Also this version seems better than the min_accuracy thing so I would just replace it. I think the reduceLR can be used all the time.

Comment (hasibzunair) [reply decavision/model_training/tfrecords_image_classifier.py:321]: Should I change the min_accuracy parameter to patience for early stopping? If I replace it, min_acc has no use, as it just adds early stopping callback.

Comment (yangobeil) [reply decavision/model_training/tfrecords_image_classifier.py:321]: Yes you replace min_accuracy by patience. You basically replace all the code that concerns min_Accuracy with your code.

Comment (hasibzunair) [reply decavision/model_training/tfrecords_image_classifier.py:321]: Alright!

Comment (hasibzunair) [reply decavision/model_training/tfrecords_image_classifier.py:321]: Made the changes.

File: decavision/model_training/tfrecords_image_classifier.py [hunk 3]
```diff
@@ -298,21 +298,32 @@ def fit(self, save_model=None, export_model=None, min_accuracy=None,
             verbose (bool): show details of training or not
             fine_tuning (bool): fine tune pretrained model or not
             l2_lambda (float): amount of L2 regularization to include in extra layer
-            min_accuracy (float): if specified, stop training when improvement in val accuracy is smaller than min_accuracy
+            patience (int): if specified, stops training when improvement in val accuracy is not observed using early stopping
```
Comment (yangobeil) [inline decavision/model_training/tfrecords_image_classifier.py]: I would rephrase this a bit. Maybe something like: if non zero, stop training when improvement in val accuracy is not observed for the given number of epochs. If used, best model is restored when training is stopped.

File: decavision/model_training/tfrecords_image_classifier.py [hunk 4]
```diff
@@ -298,21 +298,32 @@ def fit(self, save_model=None, export_model=None, min_accuracy=None,
...
             logs (str): if specified, tensorboard is used and logs are saved at this location
         """
         callbacks = None
```
Comment (yangobeil) [inline decavision/model_training/tfrecords_image_classifier.py]: I don't think this line is useful anymore since callbacks is always defined by the lines below.

File: decavision/model_training/tfrecords_image_classifier.py [hunk 5]
```diff
@@ -298,21 +298,32 @@ def fit(self, save_model=None, export_model=None, min_accuracy=None,
...
+                                                         mode='max')
+        callbacks = [reduce_lr]
+
         if logs:
             logdir = os.path.join(
                 logs, datetime.datetime.now().strftime("%Y%m%d-%H%M%S"))
             print('Fit log dir : ' + logdir)
             tensorboard_callback = tf.keras.callbacks.TensorBoard(logdir)
-            callbacks = [tensorboard_callback]
+            if not callbacks:
```
Comment (yangobeil) [inline decavision/model_training/tfrecords_image_classifier.py]: Since callbacks is always defined to include reduce_lr, I think you only need to append tensorboard here. The if is not necessary.

File: decavision/model_training/tfrecords_image_classifier.py [hunk 6]
```diff
@@ -298,21 +298,32 @@ def fit(self, save_model=None, export_model=None, min_accuracy=None,
...
+            else:
+                callbacks.append(tensorboard_callback)
 
-        # if we want to stop training when no sufficient improvement in accuracy has been achieved
-        if min_accuracy:
+        # if we want to stop training when no sufficient improvement in validation metric has been achieved
+        if patience:
             early_stop = tf.keras.callbacks.EarlyStopping(monitor='val_sparse_categorical_accuracy',
-                                                          min_delta=min_accuracy)
+                                                          patience=8,
```
Comment (yangobeil) [inline decavision/model_training/tfrecords_image_classifier.py]: patience should be set to the parameter of the function :p

File: decavision/model_training/tfrecords_image_classifier.py [hunk 7]
```diff
@@ -298,21 +298,32 @@ def fit(self, save_model=None, export_model=None, min_accuracy=None,
...
+                                                          restore_best_weights=True)
             if not callbacks:
```
Comment (yangobeil) [inline decavision/model_training/tfrecords_image_classifier.py]: Same comment as above. I think the append line is the only thing you need.

File: decavision/model_training/tfrecords_image_classifier.py [hunk 8]
```diff
@@ -400,8 +411,7 @@ def hyperparameter_optimization(self, num_iterations=20, n_random_starts=10, min
             n_random_starts (int): number of random combinations of hyperparameters first tried
             num_iterations (int): total number of hyperparameter combinations to try (aim for a 1:1 to 2:1 ratio
                 num_iterations/n_random_starts)
-            min_accuracy (float): if specified, training in single try will stop when improvement in accuracy
-                is smaller than min_accuracy
+            patience (int): if specified, stops training when improvement in val accuracy is not observed using early stopping
```
Comment (yangobeil) [inline decavision/model_training/tfrecords_image_classifier.py]: See my comment above in the fit method.

**Comment by hasibzunair** [comment]
Fixed all the changes you suggested.

**Review by yangobeil — APPROVED** [review]
I tested the code and it seems to work well. Thanks  :)
